### ISSUE TITLE
Interleaved mutual HIT

### ISSUE THREAD

**Issue by ice1000** [agda/agda#5244]
```agda
{-# OPTIONS --cubical #-}
open import Cubical.Core.Everything
interleaved mutual
 data AA : Type₀
 data BB : Type₀
 data AA where a b : AA
 data BB where g : AA → BB
 data BB where d : g a ≡ g b
 data AA where f : BB → AA
 data AA where c : f (g a) ≡ a
```

This code should check, or shouldn't?

**Comment by gallais** [comment]
The syntax of interleaved mutual blocks requires you to put the new
constructors in a `constructor` block. So this may parse but will
give probably give you a scope error saying that `AA` and `BB` have
already been declared.

Even if it were structured the right way, it probably wouldn't be
accepted with Agda as is because the grouping logic I have implemented
only targets usual inductive families (it looks at the head type constructor
of the return type of each constructor to decide which family it belongs to).

**Comment by ice1000** [comment]
@gallais The problem is that these constructors are path constructors.

```agda
{-# OPTIONS --cubical #-}
open import Cubical.Core.Everything
interleaved mutual
 data AA : Type₀
 data BB : Type₀
 constructor a b : AA
 constructor g : AA → BB
 constructor f : BB → AA
 constructor d : g a ≡ g b
 constructor c : f (g a) ≡ a
```

This code gives me the following message:

```
Could not find a matching data signature for constructor d
There was no candidate.
```

**Comment by gallais** [comment]
Yeah that's what I was saying: the grouping logic does not handle these.
You could add a special case for equality-looking return types & then go
mine the LHS & the RHS for a head constructor that already belongs to an
already declared datatype.

**Comment by ice1000** [comment]
> Yeah that's what I was saying: the grouping logic does not
> handle these.

where is this code implemented?

**Comment by ice1000** [comment]
I mean in which file or module, because I may look at it

**Comment by gallais** [comment]
https://github.com/agda/agda/blob/12ba4038629784eec568bde6d6a795e515df3e18/src/full/Agda/Syntax/Concrete/Definitions.hs#L793-L794

**Comment by ice1000** [comment]
I think changing the following code may achieve this.
https://github.com/agda/agda/blob/12ba4038629784eec568bde6d6a795e515df3e18/src/full/Agda/Syntax/Concrete/Definitions.hs#L942-L958

**Comment by ice1000** [comment]
I have finally understood what gallais meant... I don't think changing the above will make the HIT example work.

**Comment by gallais** [comment]
Nowadays the syntax has been changed to resemble the one you used
in your original message. So we should not need the `isConstructor` logic
anymore as we don't guess the family the constructor belongs to, we
expect the user to state it explicitly.

**Comment by KonjacSource** [comment]
I tried the code in origin message, and Agda told me that `g` is not in scope when checking the last line. Is that suppose to happend?

**Comment by NathanielB123** [comment]
> I tried the code in origin message, and Agda told me that `g` is not in scope when checking the last line. Is that suppose to happend?

Yeah, this is a consequence of how `interleaved mutual` is desugared, see the [documentation](https://agda.readthedocs.io/en/v2.6.3/language/mutual-recursion.html#interleaved-mutual-blocks):

> The interleaved mutual blocks get desugared into the [Forward declaration](https://agda.readthedocs.io/en/v2.6.3/language/mutual-recursion.html#mutual-recursion-forward-declaration) blocks described below by:
>
> - leaving the signatures where they are,
>
> - grouping the clauses for a function together with the first of them, and
>
> - grouping the constructors for a datatype together with the first of them.

So, all constructors for `AA` (including `c : f (g a) ≡ a`) are grouped with `data AA where a b : AA`, and `BB`'s constructors are not yet in scope.

Luckily we can transform such definitions into a form Agda accepts using a datatype of codes:
```agda
data Code : Set where
  aa : Code
  bb : Code

data El : Code → Set

AA = El aa
BB = El bb

data El where
  a b : AA
  g : AA → BB
  d : g a ≡ g b
  f : BB → AA
  c : f (g a) ≡ a
```
This transformation is called "two-sortification" in [For Generalised Algebraic Theories, Two Sorts are Enough](https://arxiv.org/pdf/2601.19426)

**Comment by KonjacSource** [comment]
That is cool, thank you so much.

**Comment by ice1000** [comment]
> This transformation is called "two-sortification" in [For Generalised Algebraic Theories, Two Sorts are Enough](https://arxiv.org/pdf/2601.19426)

Wow! I thought it was Szumi who came up with this trick.

**Comment by maxsnew** [comment]
You can also type this without changing any sorts by declaring an alias for whatever constructors you need and then patching it up after. This is how I've been handling big mutual QITs:
```agda
data AA : Type₀
data BB : Type₀
g' : AA → BB
data AA where
  a b : AA
  f : BB → AA
  c : f (g' a) ≡ a
data BB where
  g : AA → BB
  d : g a ≡ g b
g' = g
```

**Comment by NathanielB123** [comment]
> You can also type this without changing any sorts by declaring an alias for whatever constructors you need and then patching it up after.

Ah yes, that also can work, though using this trick for inductive-inductive definitions can get painful because clauses of eliminating functions also can't be interleaved. E.g.
```agda
data AA : Set
data BB : AA → Set

a' : AA
b' : BB a'

data AA where
  a : AA
  f : BB a' → AA
  g : BB (f b') → AA

data BB where
  b : BB a'
  h : (x : BB a') → BB (f x)

a' = a
b' = b

module _ (AAᴹ : AA → Set) (BBᴹ : ∀ {x} → AAᴹ x → BB x → Set)
         (aᴹ : AAᴹ a) (fᴹ : ∀ {x} → BBᴹ aᴹ x → AAᴹ (f x))
         (bᴹ : BBᴹ aᴹ b)
         (gᴹ : ∀ {x} → BBᴹ (fᴹ bᴹ) x → AAᴹ (g x))
         (hᴹ : ∀ {x} (xᴹ : BBᴹ aᴹ x) → BBᴹ (fᴹ xᴹ) (h x)) 
         where
  elimA : ∀ x → AAᴹ x
  elimB : ∀ {x} (y : BB x) → BBᴹ (elimA x) y

  elimA a     = aᴹ
  elimA (f x) = fᴹ (elimB x)
  elimA (g x) = gᴹ (elimB x) -- Error here!

  elimB b     = bᴹ
  elimB (h y) = hᴹ (elimB y)
```
is a type error because the `elimA (g x) = gᴹ (elimB x)` clause relies on `elimB b = bᴹ` (but swapping `elimA` and `elimB` also doesn't work because `elimB (h y) = hᴹ (elimB y)` relies on `elimA (f x) = fᴹ (elimB x)` - this is probably fixable by doing the same forward declaration trick for the eliminator and manually transporting in some of the clauses but... yuck).

Actually for the original HIT example, the eliminator does work out though
```agda
data AA : Set
data BB : Set
g' : AA → BB
data AA where
  a b : AA
  f : BB → AA
  c : f (g' a) ≡ a
data BB where
  g : AA → BB
  d : g a ≡ g b
g' = g

module _ (AAᴹ : AA → Set) (BBᴹ : BB → Set) 
         (aᴹ : AAᴹ a) (bᴹ : AAᴹ b) 
         (fᴹ : ∀ {x} → BBᴹ x → AAᴹ (f x))
         (gᴹ : ∀ {x} → AAᴹ x → BBᴹ (g x))
         (cᴹ : PathP (λ i → AAᴹ (c i)) (fᴹ (gᴹ aᴹ)) aᴹ)
         (dᴹ : PathP (λ i → BBᴹ (d i)) (gᴹ aᴹ) (gᴹ bᴹ)) 
         where
  elimA : ∀ x → AAᴹ x
  elimB : ∀ x → BBᴹ x
  
  elimA a     = aᴹ
  elimA b     = bᴹ
  elimA (f x) = fᴹ (elimB x)
  elimA (c i) = cᴹ i
  
  elimB (g x) = gᴹ (elimA x)
  elimB (d i) = dᴹ i
```
I find this kinda surprising honestly given `elimA (c i) = cᴹ i` really does rely on `elimB (g x) = gᴹ (elimA x)`. I guess boundary conditions are checked after all mutual clauses are defined?

**Comment by NathanielB123** [comment]
> Wow! I thought it was Szumi who came up with this trick.

Yes I think you are right: https://web.archive.org/web/20241004151846/https://lists.chalmers.se/pipermail/agda/2019/011176.html

**Comment by KonjacSource** [comment]
> is a type error because the `elimA (g x) = gᴹ (elimB x)` clause relies on `elimB b = bᴹ` (but swapping `elimA` and `elimB` also doesn't work because `elimB (h y) = hᴹ (elimB y)` relies on `elimA (f x) = fᴹ (elimB x)` - this is probably fixable by doing the same forward declaration trick for the eliminator and manually transporting in some of the clauses but... yuck).

I think function clauses should also be able to be interleaved, maybe.
I have a toy Agda-styled language that supports interleaved clauses, and it can actually check this example. [See here.](https://github.com/KonjacSource/ShiTT2/blob/main/examples/interleaved.shitt)

**Comment by ice1000** [comment]
Functions are only defined after you're done with a fully elaborated definition, which includes coverage checking and termination checking. Relying on single-clause reduction may not be a good idea.

However, you can also say that relying on intermediate clause-reduction behavior is part of the elaboration, similar to how you check HIT pattern matching. Both functions become defined after two of them are checked together, so in that sense this might be ok too

**Comment by KonjacSource** [comment]
I think two mutual functions should be considered as a single function. And it is technically ok, one can check coverage after existing mutual block, and termination check also works for partial definition.

**Comment by ice1000** [comment]
> I think two mutual functions should be considered as a single function. And it is technically ok, one can check coverage after existing mutual block, and termination check also works for partial definition.

The computational behavior is not necessarily the same as written by the clauses.

**Comment by KonjacSource** [comment]
I made a version of agda to support this XD.

https://github.com/KonjacSource/agda/blob/master/MutualExample.agda
